### PR TITLE
[dtensor] Make distribute_tensor support distributing DTensors

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124962
* #124961

Summary:
If the input tensor is a DTensor, distribute_tensor should be able to futher distribute the if the following conditions are satisfied:

1. The mesh of the DTensor and the target mesh have the same parent mesh.
2. Shard must be done from the largest dimension to the smallest dimension of the mesh.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/_tensor/api.py [hunk 1]
```diff
@@ -592,25 +592,43 @@ def distribute_tensor(
...
+                    "Currently, distribute_tensor only supports 2D parent device mesh."
+                )
+            # TODO: these checks need to be extended once we support nD parent
+            # device mesh.
+            if (
+                _mesh_resources.get_parent_mesh_dim(tensor.device_mesh) == 0
+                and next(iter(tensor.placements)).is_shard()
+                and placements[0].is_shard()
+            ):
+                raise ValueError(
```
Comment (wanchaol) [inline torch/distributed/_tensor/api.py:620]: We could support multi Shard placements on multiple mesh dim if:

The input dtensor's mesh is in the left index compare to the passed in `device_mesh` in the parent_mesh mesh_dims (In this case, we can further shard the DTensor on the same dimension)

File: torch/distributed/_tensor/api.py [hunk 2]
```diff
@@ -592,25 +592,43 @@ def distribute_tensor(
...
+                    "The DTensor is already sharded on mesh dimension 0. "
+                    "We could not shard on the other mesh dimensions."
+                )
+            local_tensor = tensor._local_tensor
```
Comment (wanchaol) [inline torch/distributed/_tensor/api.py:624]: In addition to the above checks, we also need to check if the input `dtensor` is a leaf tensor or not and make sure its grad_fn is None.

If it's not a leaf tensor, then this input DTensor already participated into some computation, which we need to throw error for such DTensors

File: test/distributed/_tensor/test_dtensor.py [hunk 3]
```diff
@@ -825,6 +825,89 @@ def test_split_tensor_1D(self) -> None:
...
+    @with_comms
+    def test_distribute_dtensor_fsdp_tp(self) -> None:
+        mesh, tensor = self._init_mesh_tensor()
+
+        # Mimic FSDP + TP, p1 == FSDP, p2 == TP
+        dtensor_tp = distribute_tensor(tensor, mesh["p2"], placements=[Shard(0)])
+        dtensor_dp_tp = distribute_tensor(
+            dtensor_tp,
+            mesh["p1"],
+            placements=[Shard(0)],
```
Comment (wanchaol) [inline test/distributed/_tensor/test_dtensor.py:861]: hmmm I think this should error out? We can only further distribute a DTensor when the mesh of the DTensor is in the left of the parent mesh, not the right.

For FSDP + TP case, we would need to enable the `strided sharding` in the `Shard` placement, @XilunWu is working on that currently

**Review by wanchaol — COMMENTED** [review]
Great addition! Have a few comments inlined :)

File: torch/distributed/_tensor/api.py [hunk 4]
```diff
@@ -592,25 +592,43 @@ def distribute_tensor(
...
+                raise ValueError(
+                    f"Cannot distribute a DTensor with placements {tensor.placements} "
+                    f"to a different placements {placements}. do you want to call "
+                    f"`redistribute` instead?"
+                )
+            return tensor
+
+        parent_mesh = _mesh_resources.get_parent_mesh(tensor.device_mesh)
+        if parent_mesh is not None and _mesh_resources.get_parent_mesh(device_mesh):
+            if parent_mesh.ndim != 2:
```
[the remark this answers is not in the document]
Comment (wz337) [reply torch/distributed/_tensor/api.py:609]: Probably need it before we re-land 2D submesh slicing?

**Comment by fegin** [comment]
Discussed with @XilunWu offline and he will have a new version with strided sharding support. Close this one in favor of his PR.

**Pull request closed without merging**